### PR TITLE
Make swr zero dependency

### PR DESCRIPTION
**Pull request by huozhi**
make swr core dependent less and light weight. the deep equal won't change much, now make as a vendor of swr lib.

* extract lite version of deep equal as lib
* remove dependency

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { dequal } from 'dequal/lite'
+import deeEqual from './libs/deep-equal'
```
Comment (shuding) [inline src/config.ts:1]:
```suggestion
import deepEqual from './libs/deep-equal'
```

File: src/config.ts [hunk 2]
```diff
@@ -64,7 +64,7 @@ const defaultConfig = {
   refreshWhenOffline: false,
   shouldRetryOnError: true,
   suspense: false,
-  compare: dequal,
+  compare: deeEqual,
```
Comment (shuding) [inline src/config.ts:67]:
```suggestion
  compare: deepEqual,
```

File: src/libs/deep-equal.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+const has = Object.prototype.hasOwnProperty
```
Comment (shuding) [inline src/libs/deep-equal.ts:1]: Let's include https://github.com/lukeed/dequal/blob/master/license as a comment at the top of this file.

**Review by shuding — CHANGES_REQUESTED** [review]

**Review by shuding — APPROVED** [review]

**Comment by lukeed** [comment]
This does more harm than good. 

Because swr is a library, and _not_ a built distribution (like a CLI), literally all this does is make the end-users' applications larger because any other deps that relied on `dequal` (or any shared dep) cannot do that anymore.

Dequal itself has no dependencies & has been extremely stable, so you had no motivation to do this.

This was done purely as a "dependency free" marketing tactic which is total bullshit & breeds distrust in library authors, like myself, who _actually_ make dependency-free packages.

**Comment by lukeed** [comment]
While this stands, as per MIT licensing, you have to include my license within your `LICENSE` as it's now part of your package. This is theft otherwise.

File: src/libs/deep-equal.ts (hunk 3 above)
Comment (developit) [reply src/libs/deep-equal.ts:6]: The license must be included in the repo root and/or readme such that it remains visible in the npm package, as npm packages are copies ([substantial portions](https://github.com/lukeed/dequal/blob/9aa73181ac7e081cd330cac67d313632ac04bb02/license#L13)) of software.

**Comment by huozhi** [comment]
@lukeed My apologies to the inappropriate way of using `dequal` here, and the misleading PR title which is not intend to marketing swr as a dependency free library. The PR purpose should be described as use dequal lite as inline module, that's it nothing else.

I'm super appreciated all the work you've done and the awesome libraries you built, feel sorry to bring such sore feeling to you because of my bad description which is definitely not anyone's purpose to harm origin author. I'll file another PR to include the license of `dequal` into swr, so much thanks for the reminding.

**Comment by lukeed** [comment]
The PR title is "Make swr zero dependency" – I'm not sure how I've been misled here.

> I'll file another PR to include the license of dequal into swr, so much thanks for the reminding

Thank you for agreeing to update the LICENSE – I'll wait for that PR – but the rest of my points still stand: inlining `dequal` does more harm than good because `swr`'s copy **cannot be shared** by/with other packages inside an application. It should be left as a dependency – because it IS a dependency.

Thank you for appreciating my work, but this does not align with those words. At best, this is _hiding_ your appreciation. And this is Vercel shipping my code as their own, which it is not: https://www.diffchecker.com/UbAYQ6Ht

**Comment by lukeed** [comment]
Thank you

**Pull request merged**